### PR TITLE
mandelbulber: init at 2.20

### PR DESCRIPTION
**Pull request by KoviRobi**
Tested on my machine (NixOS 19.03.173243.9ef3cb9b0ba, x86_64), see package request #52038

###### Motivation for this change
I wanted to generate pretty pictures.

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/graphics/mandelbulber2/default.nix
Comment (aanderse) [inline pkgs/applications/graphics/mandelbulber2/default.nix]: -> `pname = "mandelbulber2";`

File: pkgs/applications/graphics/mandelbulber2/default.nix
Comment (aanderse) [inline pkgs/applications/graphics/mandelbulber2/default.nix]: You can drop this line once you set `pname` and `version` as `name = "${pname}-${version}";` is already implied.

File: pkgs/applications/graphics/mandelbulber2/default.nix
Comment (aanderse) [inline pkgs/applications/graphics/mandelbulber2/default.nix]: `stdenv` -> `lib`.

File: pkgs/applications/graphics/mandelbulber2/default.nix
Comment (aanderse) [inline pkgs/applications/graphics/mandelbulber2/default.nix]: `stdenv.lib` -> `lib`.

File: pkgs/applications/graphics/mandelbulber2/default.nix
Comment (aanderse) [inline pkgs/applications/graphics/mandelbulber2/default.nix]: -> `meta = with lib; {`, then you can drop `stdenv.lib` usage within this attribute set.

File: pkgs/applications/graphics/mandelbulber2/default.nix
Comment (aanderse) [inline pkgs/applications/graphics/mandelbulber2/default.nix]: Please use a consistent 2 spaces per level of indentation.

**Review by aanderse — CHANGES_REQUESTED** [review]
Looking good. After a quick review I only have cosmetic points to mention. :+1:

File: pkgs/applications/graphics/mandelbulber2/default.nix
Comment (ivan) [inline pkgs/applications/graphics/mandelbulber2/default.nix]: Thank you for the packaging!

Could you also patch these three example files?

```
/nix/store/slc8qi79n23wsripnj5vgrwjr9ylb8sm-mandelbulber2-2.19# rg /usr
[...]

share/mandelbulber2/examples/gradients - reflectance and transparency.fract
14:file_background /usr/share/mandelbulber2/textures/background3.jpg;

share/mandelbulber2/examples/gradients - specular highlights.fract
12:file_background /usr/share/mandelbulber2/textures/background3.jpg;

share/mandelbulber2/examples/transparency reflectance roughness textures.fract
14:file_background /usr/share/mandelbulber2/textures/background3.jpg;
```

**Comment by ivan** [comment]
@GrahamcOfBorg build mandelbulber2

**Review by ivan — APPROVED** [review]
Built and tested on NixOS x86_64, examples load and both CPU and OpenCL rendering work. LGTM.

File: pkgs/applications/graphics/mandelbulber2/default.nix
Comment (aanderse) [inline pkgs/applications/graphics/mandelbulber2/default.nix]: @KoviRobi sorry I seem to have forgotten to ask you to provide a value for `meta.platforms`.

**Comment by KoviRobi** [comment]
While it should work on Mac and Windows as well, since I don't have either to test it I haven't added it to the platforms, my thinking was that if someone needs it on that platform they can test it and make it work.

File: pkgs/applications/graphics/mandelbulber2/default.nix
Comment (jtojnar) [inline pkgs/applications/graphics/mandelbulber2/default.nix]: Why not just use

```suggestion
    rev = version;
```

File: pkgs/applications/graphics/mandelbulber2/default.nix
Comment (jtojnar) [inline pkgs/applications/graphics/mandelbulber2/default.nix]: Can you please open an upstream issue about the hardcoded prefix and add a comment here with a link.

**Comment by KoviRobi** [comment]
Apparently there are two flags to pass to the C compiler/preprocessor, into which I want a C string, [hence](https://interest.qt-project.narkive.com/X4rxEe5W/defining-a-string-in-qmake-command-line#post10) I need to enclose the path in double-quotes. Make calls the C compiler in a shell, so I need to escape the double-quotes hence (\"). The shell calling `qmake` does backslash expansion hence (\\\"); Nix strings do backslash expansion hence (\\\\\\\").

**Review by ivan — APPROVED** [review]
Built and tested again, LGTM

File: pkgs/applications/graphics/mandelbulber2/default.nix
Comment (jtojnar) [inline pkgs/applications/graphics/mandelbulber2/default.nix]: You can use `''` strings to get rid of slash in each of `\"` and `\\`. You can also use `${placeholder "out"}` to avoid another one.
```suggestion
  qmakeFlags = [ ''DEFINES+=SHARED_DIR=\\\"${placeholder "out"}/share/mandelbulber2\\\"''
```

Comment (jtojnar) [reply pkgs/applications/graphics/mandelbulber2/default.nix]: Won’t `SHARED_PATH`, as described in https://github.com/buddhi1980/mandelbulber2/blob/ce5f5d789c78ba76ea36d0efeb79f2f2b3ca643d/mandelbulber2/deploy/README#L137 work?

```suggestion
  qmakeFlags = [ "SHARED_PATH=${placeholder ''out''}" ];
```

https://github.com/buddhi1980/mandelbulber2/blob/ce5f5d789c78ba76ea36d0efeb79f2f2b3ca643d/mandelbulber2/qmake/installer.pri#L7-L9
https://github.com/buddhi1980/mandelbulber2/blob/ce5f5d789c78ba76ea36d0efeb79f2f2b3ca643d/mandelbulber2/src/system.hpp#L48-L58

File: pkgs/applications/graphics/mandelbulber2/default.nix
Comment (jtojnar) [inline pkgs/applications/graphics/mandelbulber2/default.nix]: Could you ask upstream for this files to also respect `PREFIX` or `SHARED_PATH`?

**Comment by aanderse** [comment]
ping (triage) ... where are we at with this?

**Comment by KoviRobi** [comment]
I have folded in the latest changes from jtojnar, and have asked upstream to patch example files https://github.com/buddhi1980/mandelbulber2/issues/719 though that hasn't progressed yet. Apart from that I think that's all the comments addressed

**Comment by aanderse** [comment]
@KoviRobi Please address bot failures when you get a chance.

**Comment by KoviRobi** [comment]
Sure, though not sure how to interpret the failures, all I see is

grahamcofborg-eval
```
x86_64-linux	mandelbulber2
i686-linux	mandelbulber2
aarch64-linux	mandelbulber2
```

grahamcofborg-eval-check-meta
```
nix-env failed:
```

and it doesn't say much to me.

File: pkgs/applications/graphics/mandelbulber2/default.nix
Comment (jtojnar) [reply pkgs/applications/graphics/mandelbulber2/default.nix]: Could you add a comment linking to the upstream issue?

**Comment by KoviRobi** [comment]
Looks like the bot just needed to be started again?

**Comment by KoviRobi** [comment]
@aanderse I think this is ready to be merged? I have addressed all the comments, and the bot has managed to make everything pass.

**Comment by aanderse** [comment]
@GrahamcOfBorg build mandelbulber2

**Comment by aanderse** [comment]
@KoviRobi sorry for the delay, this PR slipped by me. Thanks for your patience and persistence!

@ivan @jtojnar any additional changes required on your part before merge?

**Comment by aanderse** [comment]
@KoviRobi any idea on whenthe next upstream release is? That release will eliminate the need for your patch here as your patch was accepted upstream...

**Comment by KoviRobi** [comment]
@aanderse I have just asked, apparently in about 2 weeks, so might be worth the wait until then

**Comment by KoviRobi** [comment]
Now using the new version, 2.20 which has the hardcoded paths patch included

File: pkgs/applications/graphics/mandelbulber2/default.nix
Comment (jtojnar) [inline pkgs/applications/graphics/mandelbulber2/default.nix]: `opencl` looks too much as a dependency, we usually name flags `enableFoo` or `withFoo` to distinguish them.

File: pkgs/applications/graphics/mandelbulber2/default.nix
Comment (jtojnar) [inline pkgs/applications/graphics/mandelbulber2/default.nix]:
```suggestion
  nativeBuildInputs = [
    qmake
    qttools
  ];

  buildInputs = [
    qtbase
    qtmultimedia
    libpng
    gsl
    libsndfile
    lzo
  ] ++ lib.optionals withOpencl [
    opencl-clhpp
    ocl-icd
  ];
```

File: pkgs/applications/graphics/mandelbulber2/default.nix
Comment (jtojnar) [inline pkgs/applications/graphics/mandelbulber2/default.nix]:
```suggestion
  qmakeFlags = [
    "SHARED_PATH=${placeholder ''out''}"
    (if opencl
     then "qmake/mandelbulber-opencl.pro"
     else "qmake/mandelbulber.pro")
  ];
```

File: pkgs/applications/graphics/mandelbulber2/default.nix
Comment (jtojnar) [inline pkgs/applications/graphics/mandelbulber2/default.nix]:
```suggestion
  sourceRoot = "${src.name}/mandelbulber2";
```

might be more reliable (e.g. when one overrides `src` to point to a local directory for testing a patch)

File: pkgs/applications/graphics/mandelbulber2/default.nix
Comment (jtojnar) [inline pkgs/applications/graphics/mandelbulber2/default.nix]:
```suggestion
{ lib
, mkDerivation
, fetchFromGitHub
, libpng
, gsl
, libsndfile
, lzo,
```

is more diff-friendly. See nixpkgs-fmt.

**Review by jtojnar — COMMENTED** [review]
Just a few finishing suggestions, mostly style suggestions for cleaner diffs, almost ready to merge.

File: pkgs/applications/graphics/mandelbulber2/default.nix
Comment (aanderse) [reply pkgs/applications/graphics/mandelbulber2/default.nix]: Or `openclSupport`.

File: pkgs/applications/graphics/mandelbulber2/default.nix
Comment (KoviRobi) [reply pkgs/applications/graphics/mandelbulber2/default.nix]: Done

File: pkgs/applications/graphics/mandelbulber2/default.nix
Comment (KoviRobi) [reply pkgs/applications/graphics/mandelbulber2/default.nix]: Done, I went for "withOpenCL" because I felt a mild preference.

File: pkgs/applications/graphics/mandelbulber2/default.nix
Comment (KoviRobi) [reply pkgs/applications/graphics/mandelbulber2/default.nix]: Done, I went with what you had, nixpkgs-fmt (which I wasn't aware of, thanks!) went for a brace-style spacing on the parentheses, e.g.
```
    (
     if opencl
     then "qmake/mandelbulber-opencl.pro"
     else "qmake/mandelbulber.pro"
    )
```
but I felt that was unnecessary so I went with what you have

File: pkgs/applications/graphics/mandelbulber2/default.nix
Comment (KoviRobi) [reply pkgs/applications/graphics/mandelbulber2/default.nix]: Done, used nixpkgs-fmt.

**Comment by KoviRobi** [comment]
Whoops, I seem to have messed up something with the commits, I just wanted to squash it into a single commit (I think that's preferred in the nixpkgs manual), let me fix that

**Comment by KoviRobi** [comment]
Sorry about the accidental mistake, I just did a `git rebase -i HEAD~5` thinking that just leaving the extra commits as above won't change anything, but it did (probably should have used `git rebase -i HEAD^5`, I just didn't pay close enough attention to the difference between `~` and `^`, now I know). This also then triggered a review request that I now don't know how to cancel, sorry ttuegel.

**Comment by jtojnar** [comment]
On the website and everywhere else, it seems to be referred as just `mandelbulber`. Would it make sense drop `2` from `pname` and attribute name? Also that is what it goes by on repology: https://repology.org/projects/?search=mandelbulber&maintainer=&category=&inrepo=&notinrepo=&repos=&families=&repos_newest=&families_newest=

**Comment by KoviRobi** [comment]
I guess so, I just got the impression that it went by mandelbulber2 for the version 2 (https://repology.org/project/mandelbulber/versions), though it's not a fork and v1 is deprecated (https://sourceforge.net/projects/mandelbulber/files/). Don't know if there was a rewrite/redesign that made or other incompatibilities that made packaging the two separately worthwhile. But anyway we don't have a v1.

**Comment by jtojnar** [comment]
Great work, thanks.

**Pull request merged**
